packages/models-library/tests/test_utils_string_substitution.py

```diff
@@ -0,0 +1,175 @@
+                - APP_HOSTNAME=some-prefix_%service_uuid%
+                - SPEAG_LICENSE_FILE=${OSPARC_ENVIRONMENT_SPEAG_LICENSE_FILE}
+                - MY_PRODUCT=$OSPARC_ENVIRONMENT_CURRENT_PRODUCT
+                - MY_EMAIL=$OSPARC_ENVIRONMENT_USER_EMAIL
+                - S4L_LITE_PRODUCT=yes
+                - AS_VOILA=1
+                - DISPLAY1=$${KEEP_DISPLAY}
+                - DISPLAY2=${SKIP_DISPLAY}
+    containers-allowed-outgoing-permit-list:
+        s4l-core:
+            - hostname: $OSPARC_ENVIRONMENT_LICENSE_SERVER_HOST
+              tcp_ports: [$OSPARC_ENVIRONMENT_LICENSE_SERVER_PRIMARY_PORT, $OSPARC_ENVIRONMENT_LICENSE_SERVER_SECONDARY_PORT]
+              dns_resolver:
+                  address: $OSPARC_ENVIRONMENT_LICENSE_DNS_RESOLVER_IP
+                  port: $OSPARC_ENVIRONMENT_LICENSE_DNS_RESOLVER_PORT
+    containers-allowed-outgoing-internet:
+        - s4l-core-stream
+    """
+
+    stringified_config = substitute_all_legacy_identifiers(stringified_config)
+
+    template = TemplateText(stringified_config)
+
+    assert template.is_valid()
+    identifiers = template.get_identifiers()
+    assert identifiers == [
+        "SIMCORE_REGISTRY",
+        "SERVICE_VERSION",
+        # NOTE: these identifier names were upgraded from legacy
+        "OSPARC_ENVIRONMENT_CONTAINER_NAME_SYM_SERVER",
+        "OSPARC_ENVIRONMENT_CONTAINER_NAME_DSISTUDIO_APP",
+        "OSPARC_ENVIRONMENT_SERVICE_UUID",
+        # -----
+        "OSPARC_ENVIRONMENT_SPEAG_LICENSE_FILE",
+        "OSPARC_ENVIRONMENT_CURRENT_PRODUCT",
+        "OSPARC_ENVIRONMENT_USER_EMAIL",
+        "SKIP_DISPLAY",
+        "OSPARC_ENVIRONMENT_LICENSE_SERVER_HOST",
+        "OSPARC_ENVIRONMENT_LICENSE_SERVER_PRIMARY_PORT",
+        "OSPARC_ENVIRONMENT_LICENSE_SERVER_SECONDARY_PORT",
+        "OSPARC_ENVIRONMENT_LICENSE_DNS_RESOLVER_IP",
+        "OSPARC_ENVIRONMENT_LICENSE_DNS_RESOLVER_PORT",
+    ]
+
+    # prepare substitutions map {id: value, ...}
+    exclude = {"SKIP_DISPLAY"}
+    substitutions = SubstitutionsDict(
+        {idr: "VALUE" for idr in identifiers if idr not in exclude}
+    )
+
+    # uses safe because we exclude some identifiers from substitution
+    resolved_stringified_config = template.safe_substitute(substitutions)
+
+    # all entries in the substitutions list were used
+    assert not substitutions.unused
+    assert substitutions.used == set(substitutions.keys())
+
+    # let's check how "VALUE"
+    assert (
+        resolved_stringified_config
+        == """
+    compose_spec:
+        service-one:
+            init: true
+            image: VALUE/simcore/services/dynamic/dy-static-file-server-dynamic-sidecar-compose-spec:VALUE
+            environment:
+                # legacy examples
+                - SYM_SERVER_HOSTNAME=VALUE
+                - APP_HOSTNAME=VALUE
+                - APP_HOSTNAME=some-prefix_VALUE
+                - SPEAG_LICENSE_FILE=VALUE
```

Comment:
I am not totally sure I understand why this one value got replaced and the one on line 123 not?
on line 122 this was an escaped dollar correct? But the one on line 123 was not.

Reply:
Line 122 did not get substituted because the "dollar" was escaped: ``$$``. ( originally ``DISPLAY1=$${KEEP_DISPLAY}``
Line 123 did not get substituted because I intentionally excluded it (se lines above)
```python
exclude = {"SKIP_DISPLAY"}
substitutions = SubstitutionsDict(
        {idr: "VALUE" for idr in template.get_identifiers() if idr not in exclude}
    )
```